release.sh

```diff
@@ -0,0 +1,164 @@
+}
+
+# get current version from Cargo.toml
+get_current_version() {
+    local crate_name="$1"
+    cargo metadata --no-deps --format-version 1 | \
+        jq -r --arg name "$crate_name" '.packages[] | select(.name == $name) | .version'
+}
+
+# Prompt user for confirmation
+confirm() {
+    local prompt="$1"
+    local response
+
+    echo -e -n "${YELLOW}${prompt} [y/N]${NC} "
+    read -r response
+
+    [[ "$response" =~ ^[Yy] ]]
+}
+
+# handle release branch workflow (CHANGELOG updates, README updates, PR to main)
+handle_release_branch() {
+    local version="$1"
+
+    log_info "Starting release preparation for version $version..."
+
+    # Update CHANGELOG and README
+    log_info "Updating CHANGELOG.md and README.md..."
+    if ! cargo release --workspace "$version" --no-publish --no-push --no-tag --execute; then
+        log_error "Failed to update CHANGELOG and README"
+    fi
+
+    if confirm "Print diff of CHANGELOG/README changes?"; then
+        git diff --stat HEAD^
+        git diff HEAD^
+    fi
+
+    if confirm "Would you like to open a PR with these changes?"; then
+        local current_branch
+        current_branch=$(git rev-parse --abbrev-ref HEAD)
+
+        log_info "Pushing changes to remote..."
+        git push origin "$current_branch"
+
+        if command -v gh >/dev/null 2>&1; then
+            gh pr create --title "release $version" --body "release $version"
+            log_success "PR created successfully"
+        else
+            log_warning "GitHub CLI not found. Please create a PR manually."
+        fi
+    fi
+}
+
+# Handle main branch workflow (publishing)
+handle_main_branch() {
+    local current_version
+    current_version=$(get_current_version)
+
+    # Check for clean working tree
+    if ! is_working_tree_clean; then
+        log_error "Working tree is not clean. Please commit or stash changes before publishing"
+    fi
+
+    # Check if version is already published
+    if is_version_published "delta_kernel"; then
+        log_error "delta_kernel at version $current_version is already published to crates.io"
+    fi
+
+    if is_version_published "delta_kernel_derive"; then
+        log_error "delta_kernel_derive version $current_version is already published to crates.io"
+    fi
+
+    log_info "Publishing delta_kernel version $current_version to crates.io..."
+    if ! cargo release; then
+        log_error "Failed to publish to crates.io"
+    fi
+    log_success "Successfully published version $current_version to crates.io"
+}
+
+# Main workflow
+main() {
+    check_requirements
+
+    if is_main_branch; then
+        if [[ $# -ne 0 ]]; then
+            log_warning "Version argument ignored on main branch - using version from Cargo.toml"
+        fi
+        handle_main_branch
+    else
+        if [[ $# -ne 1 ]]; then
+            log_error "Version argument required when on release branch\nUsage: $0 <version>"
+        fi
+        handle_release_branch "$1"
+    fi
+}
+
+# Entry point
```

Comment:
this feels... pythony. 

usually I'd just put all the stuff in `main` flat. not super opposed though